src/runtime/opencl.cpp

```diff
@@ -936,30 +956,69 @@ WEAK int halide_opencl_run(void *user_context,
 
     // Set args
     int i = 0;
+
+    // Count sub buffers needed for crops.
+    int sub_buffers_needed = 0;
+    while (arg_sizes[i] != 0) {
+        if (arg_is_buffer[i] &&
+            ((device_handle *)((halide_buffer_t *)args[i])->device)->offset != 0) {
+            sub_buffers_needed++;
+        }
+        i += 1;
+    }
+    cl_mem *sub_buffers = NULL;
+    int sub_buffers_saved = 0;
+    if (sub_buffers_needed > 0) {
+        sub_buffers = (cl_mem *)malloc(sizeof(cl_mem) * sub_buffers_needed);
+        if (sub_buffers == NULL) {
+            return halide_error_code_out_of_memory;
+        }
+        memset(sub_buffers, 0, sizeof(cl_mem) * sub_buffers_needed);
+    }
+
+    i = 0;
     while (arg_sizes[i] != 0) {
         debug(user_context) << "    clSetKernelArg " << i
                             << " " << (int)arg_sizes[i]
                             << " [" << (*((void **)args[i])) << " ...] "
                             << arg_is_buffer[i] << "\n";
         void *this_arg = args[i];
-        cl_int err;
+        cl_int err = CL_SUCCESS;
 
         if (arg_is_buffer[i]) {
             halide_assert(user_context, arg_sizes[i] == sizeof(uint64_t));
-            uint64_t opencl_handle = ((halide_buffer_t *)this_arg)->device;
-            debug(user_context) << "Mapped dev handle is: " << (void *)opencl_handle << "\n";
-            // In 32-bit mode, opencl only wants the bottom 32 bits of
-            // the handle, so use sizeof(void *) instead of
-            // arg_sizes[i] below.
-            err = clSetKernelArg(f, i, sizeof(void *), &opencl_handle);
+            cl_mem mem = ((device_handle *)((halide_buffer_t *)this_arg)->device)->mem;
+            uint64_t offset = ((device_handle *)((halide_buffer_t *)this_arg)->device)->offset;
+
+            uint64_t opencl_handle;
+            if (offset == 0) {
+                opencl_handle = (uint64_t)mem;
+            } else {
+                cl_buffer_region region = {offset, ((halide_buffer_t *)this_arg)->size_in_bytes()};
+                // The sub-buffer encompasses the linear range of addresses that
+                // span the crop.
+                mem = clCreateSubBuffer(mem, CL_MEM_READ_WRITE, CL_BUFFER_CREATE_TYPE_REGION, &region, &err);
+                sub_buffers[sub_buffers_saved++] = mem;
+                opencl_handle = (uint64_t)mem;
+            }
+            if (err == CL_SUCCESS) {
+                debug(user_context) << "Mapped dev handle is: " << (void *)opencl_handle << "\n";
+                // In 32-bit mode, opencl only wants the bottom 32 bits of
```

Comment:
Shouldn't opencl_handle be cl_mem, and then this comment would be unnecessary (it would be sizeof(cl_mem), which is also a pointer under the hood).

Reply:
I've interpreted this as "Can't opencl_handle go away now?"

Reply:
Isn't opencl_handle just a local variable, where everything assigned to it is a cl_mem?